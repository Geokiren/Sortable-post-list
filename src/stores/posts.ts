import { defineStore } from 'pinia';
import type { IPosts } from '@/types/posts';
import type { IHistory } from '@/types/history';
import { getPosts } from '@/services/postsService';
import type { AlertsType, IAlert } from '@/types/alerts';

export const usePostsStore = defineStore('posts', {
  state: () => ({
    posts: [] as IPosts,
    history: [] as IHistory,
    apiBaseUrl: import.meta.env.VITE_API_BASE_URL as string,
    pagination: {
      page: 1,
      limit: 5
    } as { page: number; limit: number },
    alerts: [] as AlertsType
  }),

  getters: {
    historyReversed(state): IHistory {
      return state.history.reverse();
    }
  },

  actions: {
    async fetchPosts() {
      if (!this.apiBaseUrl) {
        this.alerts.push({
          id: crypto?.randomUUID ? crypto.randomUUID() : Date.now(),
          type: 'error',
          message: 'API URL not set'
        });

        return;
      }

      try {
        const data = await getPosts(this.pagination);

        this.posts.push(...data);
      } catch (error) {
        this.addAlert('Could not fetch posts. Please, try again later.', 'error');
      }
    },

    addAlert(message: string, type: IAlert['type']) {
      this.alerts.push({
        id: crypto?.randomUUID ? crypto.randomUUID() : Date.now(),
        type,
        message
      });
    },

    swapIndexes(from: number, to: number) {
      const fromItem = this.posts[from];

      this.history.push({
        id: crypto?.randomUUID ? crypto.randomUUID() : Date.now(),
        message: `Moved Post ${fromItem.id} from index ${from} to index ${to}`,
        posts: this.posts.slice()
      });

      this.posts[from] = this.posts[to];
      this.posts[to] = fromItem;
    },

    timeTravel(index: number) {
      this.posts.splice(0, this.posts.length, ...this.historyReversed[index].posts);
      this.history.length = this.historyReversed.length - 1 - index;
    }
  }
});
