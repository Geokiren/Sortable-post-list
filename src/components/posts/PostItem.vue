<template>
  <div class="post-item card p-2 md:p-4 h-24 lg:h-32 mb-4 md:mb-6 flex justify-between items-center text-text-primary text-xl font-semibold" data-test="post-item">
    <div>Post {{ id }}</div>

    <div class="flex flex-col h-full" :class="buttonsContainerClass">
      <button v-if="showUp" class="caret" @click="getUp">
        <CaretSvg classes="w-5 h-5 fill-secondary" />
      </button>

      <button v-if="showDown" class="caret rotate-180" @click="getDown">
        <CaretSvg classes="w-5 h-5 fill-secondary" />
      </button>
    </div>
  </div>
</template>

<script setup lang="ts">
  import { computed } from 'vue';
  import CaretSvg from '../shared/CaretSvg.vue';
  import { usePostsStore } from '@/stores/posts';

  const props = defineProps<{
    id: number,
    index: number,
    size: number,
  }>();

  const { swapIndexes }: { swapIndexes: (from: number, to: number) => void } = usePostsStore();

  const getDown = () => {
    if (typeof props.index === 'number') {
      swapIndexes(props.index, props.index + 1);
    }
  }

  const getUp = () => {
    if (typeof props.index === 'number') {
      swapIndexes(props.index, props.index - 1);
    }
  }

  const showUp = computed(() => typeof props.index === 'number' && props.index > 0);
  const showDown = computed(() => typeof props.index === 'number' && props.index < props.size - 1);
  const buttonsContainerClass = computed(() => (showUp.value && showDown.value) ? 'justify-between' : 'justify-center');
</script>