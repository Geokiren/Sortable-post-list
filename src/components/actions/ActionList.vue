<template>
  <section id="action-list" class="flex flex-col w-full md:w-1/2 rounded-md shadow-lg h-fit bg-primary">
    <div class="flex items-center bg-white p-2 md:p-4 rounded-t-md">
      <h2 class="text-xl lg:text-3xl font-bold text-action-title p-3 tracking-wide">List of actions commited</h2>
    </div>

    <div class="flex flex-col p-3 md:p-5 lg:p-8">
      <TransitionGroup name="action-list" tag="div" v-if="!!historyReversed.length" class="transition-group-actions">
        <ActionItem v-for="(item, index) in historyReversed" :key="item.id" :historyItem="item" :isLast="isLast(index)" :index="index" />
      </TransitionGroup>

      <div v-else class="text-text-primary text-xl font-semibold text-center" data-test="no-actions">No actions commited</div>
    </div>
  </section>
</template>

<script setup lang="ts">
  import { usePostsStore } from '@/stores/posts';
  import type { IHistory } from '@/types/history';
  import ActionItem from './ActionItem.vue';
  import { computed } from 'vue';

  const { history }: { history: IHistory } = usePostsStore();

  const historyReversed = computed(() => [...history].reverse());

  const isLast = (index: number) => index === historyReversed.value.length - 1;

</script>

<style scoped>
  .action-list-enter-active,
  .action-list-leave-active {
    transition: all 0.5s;
  }

  .action-list-enter-from,
  .action-list-leave-to {
    opacity: 0;
    transform: translateX(60px);
  }
</style>