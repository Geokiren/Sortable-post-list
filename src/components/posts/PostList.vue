<template>
  <section id="post-list" class="flex flex-col gap-4 w-full md:w-1/2">
    <h2 class="text-2xl md:text-3xl font-bold text-post-title mb-2 sm:mb-5 tracking-wide">Sortable Post List</h2>
    
    <TransitionGroup name="post-list" tag="div" class="transition-group-posts" data-test="transition-group-posts">
      <PostItem v-for="(post, index) in posts" :key="post.id" :id="post.id" :index="index" :size="posts.length" />
    </TransitionGroup>
  </section>
</template>

<script setup lang="ts">
  import { usePostsStore } from '@/stores/posts';
  import type { IPosts } from '@/types/posts';
  import { onMounted } from 'vue';
  import PostItem from './PostItem.vue';

  const { posts, fetchPosts }: { posts: IPosts, fetchPosts: () => void } = usePostsStore();

  onMounted(() => {
    fetchPosts();
  });
</script>

<style scoped>
  .post-list-move {
    transition: all 0.5s;
  }
</style>