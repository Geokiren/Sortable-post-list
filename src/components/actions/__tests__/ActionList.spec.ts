import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import ActionList from '../ActionList.vue';
import { DOMWrapper } from '@vue/test-utils';
import { usePostsStore } from '@/stores/posts';
import { createPinia } from 'pinia';

describe('ActionList', () => {
  const mockData = [
    { id: 1, message: 'Action 1', posts: [] },
    { id: 2, message: 'Action 2', posts: [] },
    { id: 3, message: 'Action 3', posts: [] }
  ];

  it('Renders properly with the correct elements and text when no store data', () => {
    const store = usePostsStore(createPinia());

    const wrapper = mount(ActionList, { store });

    const tag = wrapper.element.tagName.toLowerCase();
    const titleElement: DOMWrapper<Element> = wrapper.find('h2');
    const text = titleElement.text();
    const noActionsElement: DOMWrapper<Element> = wrapper.find('[data-test="no-actions"]');

    expect(tag).toBe('section');
    expect(text).toBe('List of actions commited');
    expect(noActionsElement.exists()).toBe(true);
    expect(noActionsElement.text()).toBe('No actions commited');
  });

  it('Renders properly with the correct elements and text when having store data', () => {
    const store = usePostsStore(createPinia());

    const history = (store.$state.history = mockData);

    const wrapper = mount(ActionList, {
      store: { history },
      computed: { historyReversed: () => history.reverse() }
    });

    const actionItemElements = wrapper.findAll('[data-test="action-item"]');
    const historyReversedComputed = wrapper.vm.$options.computed.historyReversed.call(wrapper.vm);

    expect(actionItemElements.length).toBe(3);
    expect(historyReversedComputed[0].id).toBe(3);
  });

  it('Matches snapshot', () => {
    const store = usePostsStore(createPinia());

    const history = (store.$state.history = mockData);

    const wrapper = mount(ActionList, { store: { history } });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
