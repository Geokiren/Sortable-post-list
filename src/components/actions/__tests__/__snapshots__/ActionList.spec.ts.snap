// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionList > Matches snapshot 1`] = `
"<section data-v-d0f28be6="" id="action-list" class="flex flex-col w-full md:w-1/2 rounded-md shadow-lg h-fit bg-primary">
  <div data-v-d0f28be6="" class="flex items-center bg-white p-2 md:p-4 rounded-t-md">
    <h2 data-v-d0f28be6="" class="text-xl lg:text-3xl font-bold text-action-title p-3 tracking-wide">List of actions commited</h2>
  </div>
  <div data-v-d0f28be6="" class="flex flex-col p-3 md:p-5 lg:p-8">
    <transition-group-stub data-v-d0f28be6="" name="action-list" tag="div" appear="false" persisted="false" css="true" class="transition-group-actions">
      <div data-v-d0f28be6="" class="action-item card flex justify-between items-center p-2 md:p-3 lg:p-4 border-b-2" data-test="action-item">
        <div class="action-item-message text-text-primary text-xs :md:text-lg lg:text-xl font-semibold mr-2" data-test="action-item-message">Action 1</div><button class="button bg-primary-button text-primary-button-text">Time travel</button>
      </div>
      <div data-v-d0f28be6="" class="action-item card flex justify-between items-center p-2 md:p-3 lg:p-4 border-b-2" data-test="action-item">
        <div class="action-item-message text-text-primary text-xs :md:text-lg lg:text-xl font-semibold mr-2" data-test="action-item-message">Action 2</div><button class="button bg-primary-button text-primary-button-text">Time travel</button>
      </div>
      <div data-v-d0f28be6="" class="action-item card flex justify-between items-center p-2 md:p-3 lg:p-4" data-test="action-item">
        <div class="action-item-message text-text-primary text-xs :md:text-lg lg:text-xl font-semibold mr-2" data-test="action-item-message">Action 3</div><button class="button bg-primary-button text-primary-button-text">Time travel</button>
      </div>
    </transition-group-stub>
  </div>
</section>"
`;
